iree/tools/android/run_module_app/src/main.cc

```diff
@@ -0,0 +1,216 @@
+      << "registering HAL types";
+  iree_vm_instance_t* instance = nullptr;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance))
+      << "creating instance";
+
+  iree_vm_module_t* input_module = nullptr;
+  IREE_RETURN_IF_ERROR(LoadBytecodeModule(invocation.module, &input_module));
+
+  iree_hal_device_t* device = nullptr;
+  IREE_RETURN_IF_ERROR(CreateDevice(invocation.driver, &device));
+  iree_vm_module_t* hal_module = nullptr;
+  IREE_RETURN_IF_ERROR(CreateHalModule(device, &hal_module));
+
+  iree_vm_context_t* context = nullptr;
+  // Order matters. The input module will likely be dependent on the hal module.
+  std::array<iree_vm_module_t*, 2> modules = {hal_module, input_module};
+  IREE_RETURN_IF_ERROR(iree_vm_context_create_with_modules(
+      instance, modules.data(), modules.size(), iree_allocator_system(),
+      &context))
+      << "creating context";
+
+  const std::string& function_name = invocation.entry_function;
+  iree_vm_function_t function;
+  IREE_RETURN_IF_ERROR(input_module->lookup_function(
+      input_module->self, IREE_VM_FUNCTION_LINKAGE_EXPORT,
+      iree_string_view_t{function_name.data(), function_name.size()},
+      &function))
+      << "looking up function '" << function_name << "'";
+
+  IREE_RETURN_IF_ERROR(ValidateFunctionAbi(function));
+  IREE_ASSIGN_OR_RETURN(auto input_descs, ParseInputSignature(function));
+
+  absl::InlinedVector<absl::string_view, 4> input_views(
+      absl::StrSplit(invocation.inputs, '\n', absl::SkipEmpty()));
+  IREE_ASSIGN_OR_RETURN(
+      auto inputs,
+      ParseToVariantList(input_descs, iree_hal_device_allocator(device),
+                         input_views));
+
+  IREE_ASSIGN_OR_RETURN(auto output_descs, ParseOutputSignature(function));
+  vm::ref<iree_vm_list_t> outputs;
+  IREE_RETURN_IF_ERROR(iree_vm_list_create(/*element_type=*/nullptr,
+                                           output_descs.size(),
+                                           iree_allocator_system(), &outputs));
+
+  LOGI("Execute @%s", function_name.c_str());
+  IREE_RETURN_IF_ERROR(iree_vm_invoke(context, function, /*policy=*/nullptr,
+                                      inputs.get(), outputs.get(),
+                                      iree_allocator_system()))
+      << "invoking function " << function_name;
+
+  std::ostringstream oss;
+  IREE_RETURN_IF_ERROR(PrintVariantList(output_descs, outputs.get(), &oss))
+      << "printing results";
+  LOGI("Execution Result:");
+  LOGI("%s", oss.str().c_str());
+
+  inputs.reset();
+  outputs.reset();
+  iree_vm_module_release(hal_module);
+  iree_vm_module_release(input_module);
+  iree_hal_device_release(device);
+  iree_vm_context_release(context);
+  iree_vm_instance_release(instance);
+  return OkStatus();
+}
+
+void RunModuleAppMain(android_app* app) {
+  // Sleep for 2 seconds to allow tools like AGI to perform the necessary
+  // initialization.
```

Comment:
:cry:

Reply:
Yeah, I talked with AGI folks and that's what required as a reasonable workaround.